File: scripts/db/schema_hash.py
"""Utilities to hash the Supabase migration SQL files.

Used for generated type freshness checks.
"""

from __future__ import annotations

import hashlib
from pathlib import Path


def compute_migrations_hash(*, migrations_dir: Path) -> tuple[str, list[str]]:
    files = sorted(p for p in migrations_dir.glob("*.sql") if p.is_file())

    digest = hashlib.sha256()
    for file_path in files:
        digest.update(file_path.name.encode("utf-8"))
        digest.update(b"\0")
        # Normalize newlines to keep the hash stable across platforms.
        # Git working trees may contain CRLF on Windows (core.autocrlf), while the
        # repository content is typically LF.
        content = file_path.read_text(encoding="utf-8")
        content = content.replace("\r\n", "\n").replace("\r", "\n")
        digest.update(content.encode("utf-8"))
        digest.update(b"\0")

    return digest.hexdigest(), [p.name for p in files]

File: scripts/db/schema-hash.mjs
import crypto from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'

function listSqlFiles(dir) {
  const entries = fs.readdirSync(dir, { withFileTypes: true })
  const files = entries
    .filter(e => e.isFile())
    .map(e => e.name)
    .filter(name => name.endsWith('.sql'))
    .sort()

  return files.map(name => path.join(dir, name))
}

export function computeMigrationsHash({ migrationsDir }) {
  const files = listSqlFiles(migrationsDir)
  const hash = crypto.createHash('sha256')

  for (const filePath of files) {
    const content = fs.readFileSync(filePath)
    hash.update(path.basename(filePath))
    hash.update('\0')
    hash.update(content)
    hash.update('\0')
  }

  return {
    hash: hash.digest('hex'),
    files: files.map(p => path.basename(p)),
  }
}
